server/sql.go

```diff
@@ -992,3 +1011,44 @@ func isMySQLExcludeDatabase(database string) bool {
 	}
 	return true
 }
+
+func (s *Server) allowAccessDatabase(ctx context.Context, principalID int, instanceID int, databaseName string) (bool, error) {
+	database, err := s.getDatabase(ctx, instanceID, databaseName)
+	if err != nil {
+		if httpError, ok := err.(*echo.HTTPError); ok && httpError.Code == echo.ErrNotFound.Code {
+			return true, nil
+		}
+		return false, err
+	}
+
+	isDeveloper := false
+	for _, member := range database.Project.ProjectMemberList {
+		if member.PrincipalID == principalID && member.Role == string(api.Developer) {
+			isDeveloper = true
+			break
+		}
+	}
+
+	if !isDeveloper {
+		return true, nil
+	}
+
+	// Database level access control policy is the allowlist style.
```

Comment:
We should check both database and environment level policy to calculate the effective policy.